### DISCUSSION TITLE
Future of UE4SS config/settings files locations

### DISCUSSION THREAD

**Discussion by Buckminsterfullerene02** [UE4SS-RE/RE-UE4SS#476] [General]
## Changing PR: #506 

Currently we have:
- `root/CustomGameConfigs/`
- `root/MemberVarLayoutTemplates/`
- `root/UE4SS_Signatures/`
- `root/watches/`
- `root/UE4SS-settings.ini`

In PR #472 I will be adding a new file for saving your search filters. If I were to follow the current system, it would go into `root/liveview/filters.meta.json`. 

If we kept going in this direction, we would soon end up with the root directory being filled with even more random config folders than it already has, and I think that having this as clean as possible makes uninstalling UE4SS for the typical end-user easier (especially since many folders aren't completely clear if they are UE4SS or the game's).

Therefore, [this suggestion](https://github.com/UE4SS-RE/RE-UE4SS/pull/472#issuecomment-2061331808) has been made. Note the importance on keeping backwards compatibility for the existing files, so don't get your pitchforks out for that reason! It basically suggests:
- `root/UE4SS-config/liveview/` for anything within the main liveview tab
- `root/UE4SS-config/watches/` for anything to do with watches
- `root/UE4SS-config/<whatever else in the future we will have>`

Additionally we could also do these, but I think are a sub-discussion each in their own:
- `root/UE4SS-config/UE4SS-settings.ini` but the ini file name could change as well, maybe to just `settings.ini` since we'd already know that it's a UE4SS file and not a game one.
- `root/UE4SS-config/CustomGameConfigs/`
- `root/UE4SS-config/MemberVarLayoutTemplates/`
- `root/UE4SS-config/Signatures/`

Please let us know what your thoughts for this are, any other suggestions you have with config files etc.

**Comment by UE4SS** [comment]
I'm in favor of moving all files as long as it's backwards compatible on a file-by-file basis.
My opinion is that if we break mod loaders with this change (i.e. nexus), then we've done something we shouldn't have.

**Reply by mczubaj** [reply]
Hi, super late to the party, but I've noticed that for Palworld, Vortex's automatic support overwrites dwmapi.dll, which makes UE4SS use the legacy location. I'm guessing we'll need to update Vortex's Palworld extension to handle that change if we want to use the new UE4SS with it, right? Or use the override.txt file as a workaround at least (will that work though if Vortex overwrites dwmapi?).

**Reply by mczubaj** [reply]
Reason I'm asking is that we have a nice new tool for making mods, that was built on the experimental UE4SS - [PalSchema](https://github.com/Okaetsu/PalSchema)

Some people are already using it to release mods, which right now requires manually installing experimental UE4SS and manually installing the mod. [Example1](https://www.nexusmods.com/palworld/mods/2159), [example2](https://www.nexusmods.com/palworld/mods/2162).
 So it's quite possible that users will use a mod manager for all their other mods, except for PalSchema ones, leading to a bit of a mess. Their PalSchema mods won't even exist in the legacy folder.

For the record, the Vortex extension is [here](https://github.com/Nexus-Mods/game-palworld). I've been looking through it a bit, maybe I'll attempt a PR.

**Reply by Buckminsterfullerene02** [reply]
Vortex extension and any other third party tool should update to support this, as it is possible in the distant future the old, deprecated location will no longer be supported at all.

**Comment by narknon** [comment]
My current thought is that all files should go in root/ue4ss/relevantfolder/

Lowercase for Linux and Windows doesn't care. Then eventually the proxy should be changed to load ue4ss.dll from that directory, with an override available via a config file next to the proxy file (unrelated to the PR this was brought up from).

**Reply by Buckminsterfullerene02** [reply]
What do we want to call the config file? override.txt? 

Format of the config file? A relative or absolute path of the new ue4ss.dll location?

**Reply by Buckminsterfullerene02** [reply]
For example the text in the file could be one of these, and the program would react accordingly (checking for absolute path is easy):
- `C:\UE4SS\UE4SS.dll`
- `..\..\..\..\..\..\UE4SS.dll`

**Reply by UE4SS** [reply]
Name: Considering this is outside the ue4ss directory, ue4ss should be part of its name to avoid confusion. Maybe ue4ss_dll_override ?
Format: We're looking for just a path right ? So lets do whatever is easiest for mod managers to read/write.
Rel/Abs: I think both should be supported.
We may as well also require the name of the dll just in case people for some reason rename it or want to load some other dll instead.

**Comment by UE4SS** [comment]
@Buckminsterfullerene02 For the record, I think we can consider this discussion over.
We've waited a month for anyone to give their thoughts, and I think anyone can begin implementing this whenever they want to at this point.

In conclusion:
* Config location: All config files (including UE4SS-signatures) should go in `<Win64/Linux>/ue4ss/`, and sub-directories are allowed when appropriate.
* Casing: When Linux support is added, we won't be doing case-sensitive comparisons for file paths and names, so whether we go full lower case or not shouldn't matter.
* UE4SS-settings.ini rename: There were no comments on this, but in my opinion, renaming it won't give any benefit seeing as the filename is short enough and it'll just make it more annoying when people merge to this new location and the file is named differently.
* For the same reason as above, I think no files or directories should be renamed, for example UE4SS_Signatures -> Signatures.
* ue4ss.dll in the ue4ss directory: Should be just as easy to implement in a backwards compatible way as the rest of the files, and I do think it's a good idea to move this file.

**Reply by Buckminsterfullerene02** [reply]
> Config location: All config files (including UE4SS-signatures) should go in <Win64/Linux>/ue4ss/, and sub-directories are allowed when appropriate.
Casing: When Linux support is added, we won't be doing case-sensitive comparisons for file paths and names, so whether we go full lower case or not shouldn't matter.
UE4SS-settings.ini rename: There were no comments on this, but in my opinion, renaming it won't give any benefit seeing as the filename is short enough and it'll just make it more annoying when people merge to this new location and the file is named differently.
For the same reason as above, I think no files or directories should be renamed, for example UE4SS_Signatures -> Signatures.
ue4ss.dll in the ue4ss directory: Should be just as easy to implement in a backwards compatible way as the rest of the files, and I do think it's a good idea to move this file.

Agreed.

**Reply by Buckminsterfullerene02** [reply]
So just to clarify, the only files that should be left in the root directory by default (with all dumps, crashes, etc.) are as follows:
- dwmapi.dll
- crash dumps
- UE4SS.log
- Mods/

Or anything else that should still be in there or moved into root/ue4ss?

**Reply by narknon** [reply]
Imo just dwmapi.dll and whatever file we use to override dwmapi.dll for the install location.

**Comment by UE4SS** [comment]
Note that #472 added the `UE4SS-config` directory.
Luckily this hasn't made it into a release yet so we can easily change this to `ue4ss` instead as that will be a more appropriate name considering configs aren't the only thing that might go in there.

**Reply by Buckminsterfullerene02** [reply]
Yea, something I was going to change when this discussion concluded. Good to write it down though.

**Comment by Buckminsterfullerene02** [comment]
@UE4SS What do we want to do when we detect a legacy location is being used for files, such as settings file. Do we:
1. Just use the legacy location regardless of if it is also in the new location or not
2. Only use the legacy location if it is not also in the new location
3. If it is in the new location, copy the contents of the file (or directory if one) and move it to new location then use new location (probably not a good idea if mod managers are hardcoded to use legacy location, though then again this would be a major update so arguably they should be updated anyway).

**Reply by Buckminsterfullerene02** [reply]
Probably number 2 because users would install new ue4ss version and if they have duplicate copies of files, it should likely prioritise loading the new stuff over the legacy stuff due to fixes and settings changes etc.

**Reply by UE4SS** [reply]
I agree that option 2 is the best option.
Worst case scenario, mod managers that aren't updated will be overridden if the user has manually installed to the new location, which seems like a rare occurrence because how many people use a mod manager at the same time as manually messing with the files.

**Comment by Buckminsterfullerene02** [comment]
While we on the topic, do we need API.txt anymore? It's probably quite outdated by this point and there's not much reason to be using it... only reason is if you are developing offline, but who does that these days?

**Reply by UE4SS** [reply]
I don't know who, nor does it matter in my opinion, it's still important to ship some kind of documentation.
I think API.txt should be replaced in releases by the docs directory instead.
